lib/Dialect/QUIR/Transforms/BreakReset.cpp

```diff
@@ -131,13 +143,190 @@ void BreakResetPass::runOnOperation() {
+                                       const std::string &circuitName,
+                                       measureOrCallGate quantumGate) {
+  mlir::func::FuncOp mainFunc =
+      dyn_cast<mlir::func::FuncOp>(quir::getMainFunction(moduleOp));
+  assert(mainFunc && "could not find the main func");
+  mlir::OpBuilder builder(mainFunc);
+  mlir::Location location = mainFunc.getLoc();
+
+  // mangle the circuit name if there exist a circuit with the same name in
+  // input of this pass
+  auto mangledCircuitName = circuitName;
+  int cnt = 0;
+  while (inputCircuitsSymbolMap.contains(mangledCircuitName))
+    mangledCircuitName += std::to_string(cnt++);
+
+  auto circOp = builder.create<CircuitOp>(mainFunc.getLoc(), mangledCircuitName,
+                                          builder.getFunctionType(
+                                              /*inputs=*/ArrayRef<Type>(),
+                                              /*results=*/ArrayRef<Type>()));
+
+  circOp.addEntryBlock();
+  OpBuilder circuitBuilder = OpBuilder::atBlockBegin(&circOp.getBody().front());
+
+  uint argumentIndex = 0;
+  for (auto operand : quantumGate->getOperands()) {
+    circOp.insertArgument(argumentIndex, operand.getType(), {},
+                          operand.getLoc());
+    argumentIndex++;
+  }
+
+  circuitBuilder.create<mlir::quir::ReturnOp>(location, ValueRange({}));
+  return circOp;
+}
+
+void BreakResetPass::finishCircuit(mlir::quir::CircuitOp circOp) {
+  llvm::SmallVector<Type> outputTypes;
+  llvm::SmallVector<Type> inputTypes;
+  llvm::SmallVector<Value> outputValues;
+  OpBuilder circuitBuilder = OpBuilder::atBlockBegin(&circOp.getBody().front());
+
+  bool measOpFound = false;
+  circOp.walk([&](MeasureOp measOp) {
+    measOpFound = true;
+    inputTypes = TypeRange(measOp.getOperandTypes());
+    outputTypes.append(measOp.result_type_begin(), measOp.result_type_end());
+    outputValues.append(measOp.result_begin(), measOp.result_end());
+    return WalkResult::interrupt();
+  });
+
+  bool callGateOpFound = false;
+  circOp.walk([&](CallGateOp callGateOp) {
+    callGateOpFound = true;
+    inputTypes = TypeRange(callGateOp.getOperandTypes());
+    outputTypes.append(callGateOp->result_type_begin(),
+                       callGateOp->result_type_end());
+    outputValues.append(callGateOp->result_begin(), callGateOp->result_end());
+    return WalkResult::interrupt();
+  });
+
+  // sanity check that only one of measOp or callGateOp were found
+  assert(
+      ((callGateOpFound && !measOpFound) ||
+       (!callGateOpFound && measOpFound)) &&
+      "only one of measure op or call gate op can be present in input circuit");
+
+  circOp.setType(circuitBuilder.getFunctionType(
+      /*inputs=*/ArrayRef<Type>(inputTypes),
+      /*results=*/ArrayRef<Type>(outputTypes)));
+
+  circOp.walk([&](mlir::quir::ReturnOp returnOp) {
+    returnOp->insertOperands(0, ValueRange(outputValues));
+    return WalkResult::interrupt();
+  });
+}
```

Comment:
I believe this function is going to walk over the circOp 3 times:
1. Looking for Measure Op
2. Looking for callGateOp
3. Looking for returnOp.

I think it should be possible tq just walk once, and for each operation found, check if it is a measureOp and do something; check if it is a callGateOp and do something else, and finally check if it is a returnOp and do the sanity check and then do a third thing.

This should significantly reduce the amount of walking over the opertions for every circOp that gets done.